tephra-core/src/main/java/co/cask/tephra/persist/TransactionEdit.java

```diff
@@ -331,6 +367,78 @@ public void encode(TransactionEdit src, DataOutput out) throws IOException {
         }
       }
       out.writeLong(src.visibilityUpperBound);
+      // NOTE: we didn't have transaction type in V1 and V2, it was added in V3
+    }
+  }
+  
+  // TODO: refactor to avoid duplicate code among different version of codecs
+  // package-private for unit-test access
+  static class TransactionEditCodecV3 implements TransactionEditCodec {
+    @Override
+    public void decode(TransactionEdit dest, DataInput in) throws IOException {
+      if (dest.changes == null) {
+        dest.changes = Sets.newHashSet();
+      } else {
+        dest.changes.clear();
+      }
+
+      dest.writePointer = in.readLong();
+      int stateIdx = in.readInt();
+      try {
+        dest.state = TransactionEdit.State.values()[stateIdx];
```

Comment:
Using enum ordinals for serialization is brittle, as you have to be careful to only add new enum values at the end of the list.

I prefer using a short code (like a single byte) for each enum value and using that to (de)serialize.  I can describe in more detail if you want.